go/general_style_guide.md

```diff
@@ -304,3 +304,123 @@ import (
+Simple logic is not that important but sometimes it makes sense to have a simple
+test validating correct construction of bigger object so that if one forgets to
+pass configuration parameter, the validation error is caught during development
+and not in the deployment. 100% code-coverage is not something that should be
+targeted. Writing unit test also makes sense when facing a bug that could have
+been prevented by having a test for the relevant code in question.
+
+Use table-driven tests instead of separate test methods when possible. Each test
+case in table-driven test should be executed as a sub-test with
+[t.Run](https://golang.org/pkg/testing/#T.Run) as following output of `go test`
+is then easier and one can execute a single specific test case from within table
+if needed.
+
+In table-driven tests the fields of test case struct must be exported for
+improved readability.
+
+Description field must start with `case n` where `n` is the index of the test
+starting from 0.
+
+Error matchers `func(error) bool` should be used to assert errors. This makes it
+easier to match errors returned from other packages.
+
+All test validations must be included in `t.Run(...) { }` block as technically
+each test case in table is a single test.
+
+Example function to be tested:
+```go
+
+func ClusterID(obj interface{}) (string, error) {
+	customObject, ok := obj.(SpecificType)
+	if !ok {
+		return "", microerror.Maskf(wrongTypeError, "obj must be SpecificType")
+	}
+
+	return customObject.Spec.ID, nil
+}
+```
+
+Example test:
+```go
+
+func Test_ClusterID(t *testing.T) {
+	testCases := []struct {
+		Name              string
+		InputObj          interface{}
+		ExpectedClusterID string
+		ErrorMatcher      func(err error) bool
+	}{
+		{
+			Name: "case 0: clusterID returned from valid object",
+			InputObj: SpecificType{
+				Spec: Spec{
+					ID: "foobar",
+				},
+			},
+			ExpectedClusterID: "foobar",
+			ErrorMatcher:      nil,
+		},
+		{
+			Name:              "case 1: wrongTypeError returned when nil passed as obj",
+			InputObj:          nil,
+			ExpectedClusterID: "",
+			ErrorMatcher:      IsWrongTypeError,
+		},
+		{
+			Name: "case 2: wrongTypeError returned when wrong type passed as obj",
+			InputObj: struct {
+				foo string
+				bar int
+			}{
+				foo: "wrong type",
+				bar: -1,
+			},
+			ExpectedClusterID: "",
+			ErrorMatcher:      IsWrongTypeError,
+		},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.Name, func(t *testing.T) {
+			clusterID, err := ClusterID(tt.InputObj)
+
+			switch {
+			case err == nil && tt.ErrorMatcher == nil: // correct; carry on
+			case err != nil && tt.ErrorMatcher != nil:
+				if !tt.ErrorMatcher(err) {
+					t.Fatalf("error == %#v, want matching", err)
+				}
+			case err != nil && tt.ErrorMatcher == nil:
+				t.Fatalf("error == %#v, want nil", err)
+			case err == nil && tt.ErrorMatcher != nil:
+				t.Fatalf("error == nil, want non-nil")
+			}
```

Comment:
I think this can be simplified to

```go
switch {
case err == nil && tt.ErrorMatcher == nil:
	// correct; carry on
case err != nil && tt.ErrorMatcher == nil:
	t.Fatalf("error == %#v, want nil", err)
case err == nil && tt.ErrorMatcher != nil:
	t.Fatalf("error == nil, want non-nil")
case !tt.ErrorMatcher(err):
	t.Fatalf("error == %#v, want matching", err)
}
```

Reply:
:+1: